packages/map/src/lib/MapProvider.ts

```diff
@@ -488,60 +488,68 @@ export class MapProvider<StoredValue = unknown> extends JoshProvider<StoredValue
   }
 
   public [Method.Random](payload: Payload.Random<StoredValue>): Payload.Random<StoredValue> {
-    if (this.cache.size === 0) return { ...payload, data: [] };
+    const { count, unique } = payload;
 
-    const { count, duplicates } = payload;
-
-    if (this.cache.size < count) {
+    if (unique && this.cache.size < count) {
       payload.errors.push(this.error({ identifier: CommonIdentifiers.InvalidCount, method: Method.Random }));
 
       return payload;
     }
 
+    if (this.cache.size === 0) {
+      payload.errors.push(this.error({ identifier: CommonIdentifiers.MissingData, method: Method.Random, context: { unique, count } }));
+
+      return payload;
+    }
+
     payload.data = [];
 
     const keys = Array.from(this.cache.keys());
 
-    if (duplicates) {
-      while (payload.data.length < count) {
-        const key = keys[Math.floor(Math.random() * keys.length)];
-
-        payload.data.push(this.cache.get(key)!);
-      }
-    } else {
+    if (unique) {
       const randomKeys = new Set<string>();
 
       while (randomKeys.size < count) randomKeys.add(keys[Math.floor(Math.random() * keys.length)]);
 
       for (const key of randomKeys) payload.data.push(this.cache.get(key)!);
```

Comment:
Is there a way to not need 2 loops here?

Reply:
You could probably just `payload.data = randomKeys`

Reply:
Oh yeah

Reply:
I guess we could probably get the key and fetch the value in the same loop though